config/daemonconfig/mirrors.go

```diff
@@ -52,111 +64,214 @@ func makeStringSlice(slice []interface{}, cb func(string) string) ([]string, err
+	var (
+		result = hostConfig{}
+		err    error
+	)
+
+	if server != "" {
+		if !strings.HasPrefix(server, "http") {
+			server = "https://" + server
+		}
+		u, err := url.Parse(server)
+		if err != nil {
+			return hostConfig{}, fmt.Errorf("unable to parse server %v: %w", server, err)
+		}
+		result.Scheme = u.Scheme
+		result.Host = u.Host
 	}
-	dirExisted, err := file.IsDirExisted(mirrorsConfigDir)
+
+	if config.Header != nil {
+		header := http.Header{}
+		for key, ty := range config.Header {
+			switch value := ty.(type) {
+			case string:
+				header[key] = []string{value}
+			case []interface{}:
+				header[key], err = makeStringSlice(value, nil)
+				if err != nil {
+					return hostConfig{}, err
+				}
+			default:
+				return hostConfig{}, fmt.Errorf("invalid type %v for header %q", ty, key)
+			}
+		}
+		result.Header = header
+	}
+
+	result.AuthThrough = config.AuthThrough
+	result.HealthCheckInterval = config.HealthCheckInterval
+	result.FailureLimit = config.FailureLimit
+	result.PingURL = config.PingURL
+
+	return result, nil
+}
+
+func parseHostsFile(baseDir string, b []byte) ([]hostConfig, error) {
+	tree, err := toml.LoadBytes(b)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse TOML: %w", err)
+	}
+	c := struct {
+		HostFileConfig
+		// Server specifies the default server. When `host` is
+		// also specified, those hosts are tried first.
+		Server string `toml:"server"`
+		// HostConfigs store the per-host configuration
+		HostConfigs map[string]HostFileConfig `toml:"host"`
+	}{}
+
+	orderedHosts, err := getSortedHosts(tree)
 	if err != nil {
 		return nil, err
 	}
-	if !dirExisted {
-		return nil, errors.Errorf("mirrors config directory %s is not existed", mirrorsConfigDir)
+
+	var (
+		hosts []hostConfig
+	)
+
+	if err := tree.Unmarshal(&c); err != nil {
+		return nil, err
 	}
 
-	if err := filepath.Walk(mirrorsConfigDir, func(path string, info os.FileInfo, err error) error {
-		if err != nil {
-			return err
-		}
-		if info.IsDir() {
-			return nil
-		}
+	// Parse hosts array
+	for _, host := range orderedHosts {
+		config := c.HostConfigs[host]
 
-		b, err := os.ReadFile(path)
+		parsed, err := parseHostConfig(host, baseDir, config)
 		if err != nil {
-			return errors.Errorf("read mirror configuration file %s failed: %v", path, err)
+			return nil, err
 		}
+		hosts = append(hosts, parsed)
+	}
 
-		parsedMirrors, err := parseMirrorsConfig(path, b)
-		if err != nil {
-			return errors.Errorf("parse mirrors config failed: %v", err)
-		}
-		mirrors = append(mirrors, parsedMirrors...)
+	// Parse root host config and append it as the last element
+	parsed, err := parseHostConfig(c.Server, baseDir, c.HostFileConfig)
+	if err != nil {
+		return nil, err
+	}
+	hosts = append(hosts, parsed)
+
+	return hosts, nil
+}
+
+func loadHostDir(hostsDir string) ([]hostConfig, error) {
+	b, err := os.ReadFile(filepath.Join(hostsDir, "hosts.toml"))
+	if err != nil && !os.IsNotExist(err) {
```

Comment:
"hosts.toml" does not exit, should we skip parsing hosts config and don't update nydusd's mirrors config